src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -269,30 +288,83 @@ Pipe MergeTreeDataSelectExecutor::readFromParts(
-                continue;
-
-            if (minmax_idx_condition && !minmax_idx_condition->checkInHyperrectangle(
-                    part->minmax_idx.hyperrectangle, data.minmax_idx_column_types).can_be_true)
-                continue;
+        auto process_parts = [&]() {
+            auto excluded_fingerprints_map = query_context.getExcludedFingerprints()->getMap();
+            std::set<String> temp_fingerprints;
+            auto prev_parts = parts;
+            parts.clear();
 
-            if (max_block_numbers_to_read)
+            for (const auto & part : prev_parts)
             {
-                auto blocks_iterator = max_block_numbers_to_read->find(part->info.partition_id);
-                if (blocks_iterator == max_block_numbers_to_read->end() || part->info.max_block > blocks_iterator->second)
+                if (part_values.find(part->name) == part_values.end())
+                    continue;
+
+                if (part->isEmpty())
                     continue;
+
+                if (minmax_idx_condition
+                    && !minmax_idx_condition->checkInHyperrectangle(part->minmax_idx.hyperrectangle, data.minmax_idx_column_types).can_be_true)
+                    continue;
+
+                if (max_block_numbers_to_read)
+                {
+                    auto blocks_iterator = max_block_numbers_to_read->find(part->info.partition_id);
+                    if (blocks_iterator == max_block_numbers_to_read->end() || part->info.max_block > blocks_iterator->second)
+                        continue;
+                }
+
+                if unlikely(!part->fingerprint.empty())
+                {
+                    /// Skip a part if its fingerprint is meant to be excluded
+                    if unlikely (excluded_fingerprints_map.find(part->fingerprint) != excluded_fingerprints_map.end())
+                        continue;
+
+                    auto result = temp_fingerprints.insert(part->fingerprint);
+                    // found a block with the same fingerprint one the same replica, ignore it
+                    if (!result.second)
+                        continue;
+                }
+                parts.push_back(part);
+            }
+
+            if unlikely (!temp_fingerprints.empty())
+            {
+                auto duplicates = query_context.getFingerprints()->add(Strings{temp_fingerprints.begin(),temp_fingerprints.end()});
+                if (!duplicates.empty())
+                {
+                    // if any duplicates appeared during the first pass, adding them to the exclusion, so they will be skipped on second pass
+                    query_context.getExcludedFingerprints()->add(duplicates);
+                    throw Exception("Found duplicate fingerprints while processing query.", ErrorCodes::DUPLICATED_FINGERPRINTS);
+                }
             }
+        };
 
-            parts.push_back(part);
+        /// TODO(xjewer): should we add a setting on max retries?
```

Comment:
Not clear to me, but how we can get more than one retry here?

Reply:
instead of a boolean `retry` have some counter 🤷‍♂️